File: app/src/main/java/com/cybersiren/android/protocol/MessagePadding.kt
package com.cybersiren.android.protocol

import java.security.SecureRandom

object MessagePadding {

    private val blockSizes = listOf(256, 512, 1024, 2048)

    fun optimalBlockSize(dataSize: Int): Int {

        val totalSize = dataSize + 16

        for (blockSize in blockSizes) {
            if (totalSize <= blockSize) {
                return blockSize
            }
        }

        return dataSize
    }

    fun pad(data: ByteArray, targetSize: Int): ByteArray {
        if (data.size >= targetSize) return data

        val paddingNeeded = targetSize - data.size

        if (paddingNeeded <= 0 || paddingNeeded > 255) return data

        val result = ByteArray(targetSize)

        System.arraycopy(data, 0, result, 0, data.size)

        for (i in data.size until targetSize) {
            result[i] = paddingNeeded.toByte()
        }

        return result
    }

    fun unpad(data: ByteArray): ByteArray {
        if (data.isEmpty()) return data

        val last = data[data.size - 1]
        val paddingLength = last.toInt() and 0xFF

        if (paddingLength <= 0 || paddingLength > data.size) return data

        val start = data.size - paddingLength
        for (i in start until data.size) {
            if (data[i] != last) {
                return data
            }
        }

        return data.copyOfRange(0, start)
    }
}

File: ios/bitchat/localPackages/BitFoundation/Sources/BitFoundation/MessagePadding.swift
import struct Foundation.Data

struct MessagePadding {

    static let blockSizes = [256, 512, 1024, 2048]

    static func pad(_ data: Data, toSize targetSize: Int) -> Data {
        guard data.count < targetSize else { return data }

        let paddingNeeded = targetSize - data.count

        guard paddingNeeded > 0 && paddingNeeded <= 255 else { return data }

        var padded = data

        padded.append(contentsOf: Array(repeating: UInt8(paddingNeeded), count: paddingNeeded))
        return padded
    }

    static func unpad(_ data: Data) -> Data {
        guard !data.isEmpty else { return data }
        let last = data.last!
        let paddingLength = Int(last)

        guard paddingLength > 0 && paddingLength <= data.count else { return data }

        let start = data.count - paddingLength
        let tail = data[start...]
        for b in tail { if b != last { return data } }
        return Data(data[..<start])
    }

    static func optimalBlockSize(for dataSize: Int) -> Int {

        let totalSize = dataSize + 16

        for blockSize in blockSizes {
            if totalSize <= blockSize {
                return blockSize
            }
        }

        return dataSize
    }
}
